tests/unit/library/test_json.py

```diff
@@ -174,37 +174,89 @@ def test_bad_number(self, bad_string, good_bytes, failure_byte, allowed_bytes):
-    schema = """{ "type": "string" }"""
+class TestString:
+    @pytest.mark.parametrize(
+        "my_string",
+        [
+            "with_underscore",
+            "ALLCAPS",
+            "with a space",
+            "MiXeD cAsInG",
+            "with-hyphen",
+            "Mix case_underscore-hyphens",
+            "with a comma, in the string",
+            "A full stop.",
+            """How about
+                a
+                multiline string?""",
+            "A \t tab \t between \t words",
+            r"End with backslash \ ",
+            "Have a forward / slash",
+            "Include [the] odd {brace} and (parentheses)",
+            "Some more symbols: ; are useful!",
+        ],
+    )
+    @pytest.mark.parametrize("temperature", [None, 0.1, 1])
+    def test_smoke(self, my_string: str, temperature):
+        schema = """{ "type": "string" }"""
 
-    # First sanity check what we're setting up
-    schema_obj = json.loads(schema)
-    validate(instance=my_string, schema=schema_obj)
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=my_string, schema=schema_obj)
 
-    # The actual check
-    generate_and_check(my_string, schema_obj, desired_temperature=temperature)
+        # The actual check
+        generate_and_check(my_string, schema_obj, desired_temperature=temperature)
+
+    @pytest.mark.parametrize("my_string", ["aA", "aB", "aK", "aZ"])
+    def test_regex(self, my_string: str):
+        schema = """{ "type": "string", "pattern": "a[A-Z]"}"""
+
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=my_string, schema=schema_obj)
+
+        # The actual check
+        generate_and_check(my_string, schema_obj)
+
+    @pytest.mark.parametrize(
+        ["bad_string", "good_bytes", "failure_byte", "allowed_bytes"], [("aB", b"", b"a", {b'"'})]
+    )
+    def test_regex_bad(self, bad_string: str, good_bytes, failure_byte, allowed_bytes):
+        schema = """{ "type": "string", "pattern": "a[A-Z]"}"""
+        schema_obj = json.loads(schema)
+        check_match_failure(
+            bad_string=bad_string,
+            good_bytes=good_bytes,
+            failure_byte=failure_byte,
+            allowed_bytes=allowed_bytes,
+            schema_obj=schema_obj,
+        )
+
+    @pytest.mark.parametrize("my_string", ["a", "bb", "ccc"])
+    def test_min_and_maxLength(self, my_string: str):
+        schema = """{ "type": "string", "minLength": 1, "maxLength": 3}"""
+
+        # First sanity check what we're setting up
+        schema_obj = json.loads(schema)
+        validate(instance=my_string, schema=schema_obj)
+
+        # The actual check
+        generate_and_check(my_string, schema_obj)
+
+    @pytest.mark.parametrize(
+        ["bad_string", "good_bytes", "failure_byte", "allowed_bytes"],
+        [("bb", b"bb", b"a", {b'"'}), ("dddd", b"ddd", b"d", {b'"'})],
```

Comment:
Should this be `[("bb", b"", b"b", {b'"'}), ("dddd", b"", b"d", {b'"'})]` ?

Reply:
The first one is the problem. According to the schema, that should pass, so we should be getting a 'ParserException _not_ thrown' error from pytest. But that's not what's happening, and I need to figure out why.

Reply:
Reason for the missing `ParserExceptions` is that I was forgetting to include the double quotes in the string I fed to the grammar.